nova/scheduler/external.py

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2025 SAP SE
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+"""
+This module provides functionality to interact with an external scheduler API
+to reorder and filter hosts based on additional criteria.
+
+The external scheduler API is expected to take a list of weighed hosts and
+their weights, along with the request specification, and return a reordered
+and filtered list of host names.
+"""
+from oslo_log import log as logging
+import requests
+
+from nova.scheduler import utils
+
+LOG = logging.getLogger(__name__)
+
+
+def call_external_scheduler_api(weighed_hosts, weights, spec_obj, url):
+    """Reorder and filter hosts using an external scheduler service."""
+    if not url or not weighed_hosts:
+        return weighed_hosts
+    json_data = {
+        "spec": spec_obj.obj_to_primitive(),
+        "rebuild": utils.request_is_rebuild(spec_obj),
+        # Only provide basic information for the hosts for now.
+        # The external scheduler is expected to fetch statistics
+        # about the hosts separately, so we don't need to pass
+        # them here.
+        "hosts": [
+            {
+                "host": h.host,  # e.g. nova-compute-bb123
+                "hypervisor_hostname": h.hypervisor_hostname,
+                "status": h.status,
+            } for h in weighed_hosts
+        ],
+        # Also pass previous weights from the Nova weigher pipeline.
+        # The external scheduler api is expected to take these weights
+        # into account if provided.
+        "weights": weights,
+    }
+    LOG.debug("Calling external scheduler API with %s", json_data)
+    try:
+        response = requests.post(url, json=json_data, timeout=10)
```

Comment:
Should we make the timeout configurable?

Reply:
The problem is: if we don't make it configurable right now and default to 10s, we will have to do a code change once 10s don't cut it (in a big region) anymore. Hence, any value we expect could maybe change should be configurable from the start.

Reply:
I mean should be pretty straightforward to do. I'll include it in a bit.